Leverage the udev db

tasleson opened the pull request:
With the addition of Stratis support to libblkid, the udev db has
the basic information needed for stratisd.  This change leverages
the udev db where appropriate, to limit when we go to disk to
read up the meta data.

Signed-off-by: Tony Asleson <tasleson@redhat.com>

agrover commented:
First I have a very basic question I need help understanding: why don't we ever need to call settle in non-test code?

Beyond that, I'd love to see udev settle not require a cmd, if not now then in the future. 

The other big part of this appears to be a move from the DevOwnership struct to a Usage struct, for which having a little more explanation and rationale for would be helpful too.

tasleson commented:
@agrover 

>First I have a very basic question I need help understanding: why don't we ever need to call settle in non-test code?

Good question, in general we will need too if we do a change and then need to see it reflected in the udev db.

> Beyond that, I'd love to see udev settle not require a cmd, if not now then in the future.

I'll look into this.

> The other big part of this appears to be a move from the DevOwnership struct to a Usage struct, for which having a little more explanation and rationale for would be helpful too.

No real reason, while I was re-factoring I created a new struct while working out the old usage changes.  If you like I could rename it back to it's original name.  The values have changes slightly as I removed the `pool_uuid` and `device_uuid` from `ours` and added a string to `theirs`.

tasleson commented:
@agrover 

Snippit of udevadm settle
```c
        deadline = now(CLOCK_MONOTONIC) + timeout * USEC_PER_SEC;

        /* guarantee that the udev daemon isn't pre-processing */
        if (getuid() == 0) {
                struct udev_ctrl *uctrl;

                uctrl = udev_ctrl_new(udev);
                if (uctrl != NULL) {
                        if (udev_ctrl_send_ping(uctrl, MAX(5U, timeout)) < 0) {
                                log_debug("no connection to daemon");
                                udev_ctrl_unref(uctrl);
                                return EXIT_SUCCESS;
                        }
                        udev_ctrl_unref(uctrl);
                }
        }

        queue = udev_queue_new(udev);
        if (!queue) {
                log_error("unable to get udev queue");
                return EXIT_FAILURE;
        }

        pfd[0].events = POLLIN;
        pfd[0].fd = udev_queue_get_fd(queue);
        if (pfd[0].fd < 0) {
                log_debug("queue is empty, nothing to watch");
                rc = EXIT_SUCCESS;
                goto out;
        }

        for (;;) {
                if (exists && access(exists, F_OK) >= 0) {
                        rc = EXIT_SUCCESS;
                        break;
                }

                /* exit if queue is empty */
                if (udev_queue_get_queue_is_empty(queue)) {
                        rc = EXIT_SUCCESS;
                        break;
                }

                if (now(CLOCK_MONOTONIC) >= deadline)
                        break;

                /* wake up when queue is empty */
                if (poll(pfd, 1, MSEC_PER_SEC) > 0 && pfd[0].revents & POLLIN)
                        udev_queue_flush(queue);
        }

out:
        udev_queue_unref(queue);
        return rc;
```

The C library exports the `queue` functions, but not the `cntrl` functions.  Furthermore, as stated here: https://www.freedesktop.org/software/systemd/man/libudev.html 

*"... libudev also exports legacy APIs that should not be used by new software (and as such are not documented as part of this manual). This includes the hardware database known as udev_hwdb (please use the new sd-hwdb(3) API instead) and the **udev_queue** object to query the udev daemon (which should not be used by new software at all).*

So even though we could get close to replicating the `settle` functionality, unless we duplicate `cntrl` interface it wouldn't be the same, but according to the docs, the queue functions should be avoided in new software.

I think ideally a settle function should be added to the C library which could be utilized by the `udevadm` command line tool, but I'm not sure how well received that would be.

tasleson commented:
Re-based

tasleson commented:
Addresses https://github.com/stratis-storage/stratisd/issues/822

mulkieran commented:
I'm changing this to review-prelim, because it will definitely need to be squashed, at least.

mulkieran commented on src/bin/stratisd.rs:
```diff
@@ -7,16 +7,16 @@
 extern crate clap;
 #[cfg(feature = "dbus_enabled")]
 extern crate dbus;
+extern crate devicemapper;
```
All this is just a change of ordering....perhaps rustfmt did it. I would think that either it is a mistake or it belongs in a separate PR.

mati865 replied:
New rustfmt does sort `extern`, `use` and some other things.

tasleson replied:
Those are `rustfmt` changes

mulkieran replied:
That's what I figured, somehow. But I don't think it belongs in this PR if it is just a re-ordering.

agrover replied:
What happened here? Did our mass rustfmt conversion PR miss this file?

Can we just merge this and get on with it?

tasleson replied:
I think what happened is a difference in rustfmt versions.  This PR has spanned the change in versions and I think an older version did this as I reverted this change and the current rustfmt did not mess with it

tasleson commented:
@mulkieran In a previous comment you requested I squash, all the changes were individual commits before.

mulkieran commented:
I'm sorry you misunderstood me. All I really meant is that with those typo fixes and miscellaneous bits at the end, it really shouldn't be designated as review-final.

tasleson commented:
I've rebased and removed the binary handling stuff I originally had in here.

tasleson commented:
Re-based

tasleson commented:
I would really like to know how what's in `master` branch ran CI clean with the same test code that is somehow  failing to compile today.

Working: https://travis-ci.org/stratis-storage/stratisd/jobs/393085873

Failing today:
```
error[E0599]: no method named `write_all` found for type `std::io::Cursor<std::vec::Vec<u8>>` in the current scope
    --> src/engine/strat_engine/backstore/metadata.rs:1237:23
     |
1237 |             buf_newer.write_all(&sector).unwrap();
     |                       ^^^^^^^^^
     |
     = help: items from traits can only be used if the trait is in scope
     = note: the following trait is implemented but not in scope, perhaps add a `use` for it:
             candidate #1: `use std::io::Write;`
error: aborting due to previous error

```

mati865 commented:
Looks pretty odd but should build after you add `use std::io::Write;` to `src/engine/strat_engine/backstore/metadata.rs`.

tasleson commented:
@mati865 Thanks!  I'm not sure, but for some reason when I did the re-base, the code went from (`metadata.rs`):

```rust
#[cfg(test)]
mod tests {
    use std::io::{Cursor, Write};
```
to

```rust
#[cfg(test)]
mod tests {
    use std::io::Cursor;
```

agrover commented on src/engine/strat_engine/backstore/device.rs:
```diff
@@ -43,3 +46,159 @@ pub fn resolve_devices<'a>(paths: &'a [&Path]) -> StratisResult<HashMap<Device,
+    } else {
+        true
+    }
+}
```
another option:
```
    !((device.contains_key("ID_PART_TABLE_TYPE") && !device.contains_key("ID_PART_ENTRY_DISK"))
        || device.contains_key("ID_FS_USAGE"))
```
might be better? OR go the other way with:
```
    if device.contains_key("ID_PART_TABLE_TYPE") {
        if device.contains_key("ID_PART_ENTRY_DISK") {
            true
        } else {
            false
        }
    } else if device.contains_key("ID_FS_USAGE") {
        false
    } else {
        true
    }
```
might actually be best. Doing a little of both seems odd, though.

mulkieran replied:
I prefer first suggested option.

tasleson replied:
Done

agrover commented on src/engine/strat_engine/backstore/device.rs:
```diff
@@ -43,3 +46,159 @@ pub fn resolve_devices<'a>(paths: &'a [&Path]) -> StratisResult<HashMap<Device,
+            String::from(format!(
+                "We expected to find the block device {:?} \
+                 in the udev db",
+                devnode
```
With our current formatting, I think this string can all fit on one line.

tasleson replied:
Done

tasleson commented:
Superseded by https://github.com/stratis-storage/stratisd/pull/999
